keps/sig-node/20200303-shared-pod-limits.md

```diff
@@ -0,0 +1,326 @@
+  - image: proxy
+    name: envoy:latest
+  - name: nginx
+    image: nginx:latest
+    command: [ "/usr/bin/tail", "-f", "/dev/null"] 
+    resources:
+      requests:
+        memory: 128M
+        cpu: "0.5"
+      limits:
+        memory: 256M
+        cpu: "1"
+```
+
+This pod belongs to the `Burstable` QoS level because it defines multiple containers, of which only one has resource limits, and there is a 
+difference between the `Request` and the `Limit` for both these resources. Effectively, the containers which don't define any limits are
+not constrained in any way. They are provided the same level of resources as a container in a `BestEffort` pod would, except that they belong to a pod with a higher priority, so they are more proofed from causing evictions.
+
+The cgroup hierarchy for each of these resources (memory and cpu) would be this:
+
+<pre>
+  QoS CGroup (one of guaranteed, burstable, or besteffort)
+   |
+   \ pod (memory: unlimited, cpu: unlimited)
+      |
+      +-- container0 (pause container, memory: unlimited, cpu: unlimited)
+      |
+      +-- container1 (shell container, memory: unlimited, cpu: unlimited)
+      |
+      +-- container2 (proxy container, memory: unlimited, CPU: unlimited)
+      |
+      +-- container3 (nginx container, memory: 256M limit, CPU: 1 core)
+</pre>
+
+By setting the `ShareBurstableLimits` attribute on the Pod spec to `true`, the following cgroup hierarchy would be configured:
+
+<pre>
+  QoS CGroup (one of guaranteed, burstable, or besteffort)
+   |
+   \ pod (memory: 256M limit, CPU: 1 core)
+      |
+      +-- container0 (pause container, memory: unlimited, cpu: unlimited)
+      |
+      +-- container1 (shell container, memory: unlimited, cpu: unlimited)
+      |
+      +-- container2 (proxy container, memory: unlimited, CPU: unlimited)
+      |
+      +-- container3 (nginx container, memory: 256M limit, CPU: 1 core)
+</pre>
+
+This has the following effect:
+1. No change will be noticed for the nginx container
+1. The shell and proxy containers will be limited to the amount of resources specified on the Pod cgroup level - no more than 256M memory and 1 CPU core.
+1. The pause container will not be affected since it doesn't use any resources anyways.
+1. The total resource usage for this Pod will be more predictable as far as the Kubelet is concerned, since the shell container can't consume an unlimited amount of resources.
+
+Effectively, if the shell isn't being used, all of the currently unused resources which are allowed to be consumed would be usable also in the proxy container.
+
+In this example the pod memory limit is set by manipulating the limit on a single container. There is still the possibility to constrain usage for any specific container by specifying the limits in that container's section. For example, consider the following Pod specification:
+
+```yaml
+apiVersion: v1
+kind: Pod
+metadata:
+  labels:
+    run: nginx
+  name: nginx
+spec:
+  shareBurstableLimits: true
```

Comment:
could a runtime class basically specify this bounding behavior rather than complicating the user api.

Reply:
How would a runtime class apply here? My understanding of the `RuntimeClass` object is that it affects the scheduler's decision to assign a pod to a node, whereas this feature is about setting the Linux controlled pod-level cgroups so that runtime enforcement could be done on the pod level.

Also, consider a use-case where the user has two pods that should be run with the same underlying runtime (i.e. the standard docker runtime, even if it's using Containerd). One pod needs this feature, the other wants to have the current `Burstable` behavior where unconstrained pods are really unconstrained. 

If this functionality is tied to the `RuntimeClass`, then the significance would be that the developer would need to be granted privileges to duplicate the default `RuntimeClass` in order to generate one where this feature is enabled, and then change the deployment to use it. Seems to me that this would complicate matters for users trying to use the functionality.

Reply:
Instead of this boolean, should we think about specifying resources at the pod level and then allow containers to set their own values as long as they don't exceed the pod?

Reply:
That's also a possibility. If we go this path, we might not even need the validation that the containers don't exceed the pod - the Linux cgroups controllers will do the right thing even if the leaf cgroup limit exceed the parent cgroup limit.

The disadvantage to specifying the resources on the pod level directly is that it's an established pattern in Kubernetes to have admission controllers modify pods and add containers transparently, for all sorts of use-cases (Istio, etc...). This means that:
1. It would not be clear from looking at a deployment YAML what the correct amount of resources would be
1. Since different versions of admission controllers like Istio could add the same container with different resource requirements, it makes the entire pod more fragile. Upgrading one component in the system would cause hard-to-understand failures across the system.
1. It complicates the VPA mechanism. In the proposed approach the VPA needs to only be configured to watch a specific container and scale just it. If we define the resources on the pod level, the VPA needs to work a lot harder to scale both the set of containers and the pod itself.

I proposed this boolean because setting the resources on the pod level means that we would need to do additional validations on the containers in order to not exceed the pod level - so basically making it possible for the user to specify something that doesn't make sense. The boolean approach can never cause the user to provide invalid input, so it's (at least to me) a little easier to understand what will happen based on what is defined. It also transparently works with pods that have additional containers added to them behind the scenes.